python/plugins/processing/algs/qgis/TilesXYZ.py

```diff
@@ -172,103 +177,132 @@ def prepareAlgorithm(self, parameters, context, feedback):
         self.layers = [l for l in project.layerTreeRoot().layerOrder() if l in visible_layers]
         return True
 
+    def render_single_tile(self, metatile):
+        if self.feedback.isCanceled():
+            return
+            # Haven't found a better way to break than to make all the new threads return instantly
+
+        if "Dummy" in threading.current_thread().name: # single thread testing
+            specific_settings = list(self.settings_dict.values())[0]
+        else:
+            try:
+                # guess we're not the only one using TPE in QGIS, cannot assume 0_#, it's sometimes 3 or 4_#
+                specific_settings = self.settings_dict[threading.current_thread().name[-1]] # last number only
+            except:
+                print("Exception! our threads don't match with our settings! ")
+
+        size = QSize(self.tile_width * metatile.rows(), self.tile_height * metatile.columns())
+        self.extent = QgsRectangle(*metatile.extent())
+        specific_settings.setExtent(self.wgs_to_dest.transformBoundingBox(self.extent))
+        specific_settings.setOutputSize(size)
+
+        image = QImage(size, QImage.Format_ARGB32_Premultiplied)
+        image.fill(Qt.transparent)
+        dpm = specific_settings.outputDpi() / 25.4 * 1000
+        image.setDotsPerMeterX(dpm)
+        image.setDotsPerMeterY(dpm)
+        painter = QPainter(image)
+        job = QgsMapRendererCustomPainterJob(specific_settings, painter)
+        job.renderSynchronously()
+        painter.end()
+
+        for r, c, tile in metatile.tiles:
+            tile_img = image.copy(self.tile_width * r, self.tile_height * c, self.tile_width, self.tile_height)
+            self.writer.write_tile(tile, tile_img)
+
+        self.progress += 1
+        self.feedback.setProgress(100 * (self.progress / self.metatiles_total))
```

Comment:
did you explored teh use of QgsTask? in this case not sure the result of value progress when it's shared among concurrent threads

Reply:
I didn't not look at QgsTask

I didn't notice any thread problems with writer, but even if progress has threading alignment issues, and is +-1% along the way, it's only the progress UI indicator, so I'm not sure anyone would even notice.

I already checked the tile batches I made and didn't see any wrong tiles so I think writer is thread safe, I'd better double check.

Reply:
QgsTask allow stop resume, integration with task progress bar, mechanism to separate process with main thread UI iteration. BTW I did use it extensively to be sure of all side effects using different concurrent model. @nyalldawson can, for sure give a better, vision.